editor/src/clj/editor/editor_extensions/vm.clj

```diff
@@ -0,0 +1,204 @@
+           [org.luaj.vm2.compiler LuaC]))
+
+(set! *warn-on-reflection* true)
+
+(defn read
+  "Read a string with a chunk of lua code and return a Prototype for bind"
+  ^Prototype [chunk chunk-name]
+  (.compile LuaC/instance
+            (ByteArrayInputStream. (.getBytes ^String chunk StandardCharsets/UTF_8))
+            chunk-name))
+
+(deftype LuaVM [env lock])
+
+(defn ^Globals env
+  "Return VM environment (Globals)
+
+  Returned value is not thread safe"
+  [^LuaVM vm]
+  (.-env vm))
+
+(defn ^ReentrantLock lock
+  "Return lock used for synchronising access to this VM"
+  [^LuaVM vm]
+  (.-lock vm))
+
+(defn bind
+  "Bind a prototype into the VM, returns a 0-arg LuaFunction that evaluates the
+  prototype code when invoked"
+  [prototype vm]
+  (LuaClosure. prototype (env vm)))
+
+(defn must-lock?
+  "Checks if the current thread must lock the VM when interacting with it"
+  []
+  (not (.startsWith (.getName (Thread/currentThread)) "Coroutine-")))
+
+(defmacro with-lock
+  "Convenience macro for interacting with a Lua VM under a lock"
+  [vm & body]
+  `(if (must-lock?)
+     (let [lock# (lock ~vm)]
+       (try
+         (.lock lock#)
+         (do ~@body)
+         (finally (.unlock lock#))))
+     (do ~@body)))
+
+(defn make
+  "Create a fresh Lua VM
+
+  See org.luaj.vm2.lib.jse.JsePlatform/standardGlobals for typical setup"
+  []
+  (LuaVM. (Globals.) (ReentrantLock.)))
+
+(defn- invoke ^Varargs [vm ^LuaFunction lua-fn args]
+  (with-lock vm (.invoke lua-fn (LuaValue/varargsOf (into-array LuaValue args)))))
+
+(defn invoke-1
+  "Call a lua function while holding a lock on the VM
+
+  Return a LuaValue, the first value returned by the function"
+  ^LuaValue [vm lua-fn & lua-args]
+  (.arg1 (invoke vm lua-fn lua-args)))
+
+(defn invoke-all
+  "Call a lua function while holding a lock on the VM
+  Return a vector of all returned LuaValues"
+  [vm ^LuaFunction lua-fn & lua-args]
+  (let [varargs (invoke vm lua-fn lua-args)
+        n (.narg varargs)]
+    (loop [acc (transient [])
+           i 0]
+      (if (= n i)
+        (persistent! acc)
+        (let [next-i (inc i)]
+          (recur (conj! acc (.arg varargs next-i)) next-i))))))
+
+(defn wrap-userdata
+  "Wraps the value into a LuaUserdata"
+  [x]
+  (LuaValue/userdataOf x))
+
+(defn unwrap-userdata
+  "Returns the wrapped value"
+  [^LuaUserdata userdata]
+  (.userdata userdata))
+
+(defprotocol ->Lua
+  (->lua [x]))
+
+(defprotocol ->Clj
+  (->clj [lua-value vm]))
+
+(extend-protocol ->Lua
+  nil (->lua [_] LuaValue/NIL)
+  Object (->lua [x] (LuaValue/userdataOf x))
+  LuaValue (->lua [x] x)
+  Number (->lua [x] (LuaValue/valueOf (double x)))
+  Boolean (->lua [x] (LuaValue/valueOf x))
+  String (->lua [x] (LuaValue/valueOf x))
+  Named (->lua [x] (LuaValue/valueOf (.getName x)))
+  List (->lua [x] (LuaValue/listOf (into-array LuaValue (mapv ->lua x))))
+  Map (->lua [x] (LuaValue/tableOf (into-array LuaValue (into [] (comp cat (map ->lua)) x)))))
```

Comment:
Yes, if we pass a record here, it will be serialized as a table. But we generally want to reduce the serialization of data structures to a minimum, so the solution here is to not pass records to the Lua runtime.